service/src/main/java/bio/terra/workspace/service/resource/controlled/flight/create/CreateAzureIpStep.java

```diff
@@ -0,0 +1,117 @@
+package bio.terra.workspace.service.resource.controlled.flight.create;
+
+import static bio.terra.workspace.service.workspace.flight.WorkspaceFlightMapKeys.ControlledResourceKeys.CREATION_PARAMETERS;
+
+import bio.terra.cloudres.azure.resourcemanager.common.Defaults;
+import bio.terra.cloudres.azure.resourcemanager.compute.ComputeManagerCow;
+import bio.terra.cloudres.azure.resourcemanager.compute.data.CreatePublicIpRequestData;
+import bio.terra.stairway.FlightContext;
+import bio.terra.stairway.FlightMap;
+import bio.terra.stairway.Step;
+import bio.terra.stairway.StepResult;
+import bio.terra.stairway.exception.RetryException;
+import bio.terra.workspace.app.configuration.external.AzureConfiguration;
+import bio.terra.workspace.generated.model.ApiAzureIpCreationParameters;
+import bio.terra.workspace.service.crl.CrlService;
+import bio.terra.workspace.service.resource.controlled.ControlledAzureIpResource;
+import bio.terra.workspace.service.workspace.WorkspaceService;
+import bio.terra.workspace.service.workspace.flight.WorkspaceFlightMapKeys;
+import bio.terra.workspace.service.workspace.model.AzureCloudContext;
+import com.azure.resourcemanager.network.models.PublicIpAddress;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class CreateAzureIpStep implements Step {
+
+  private static final Logger logger = LoggerFactory.getLogger(CreateGcsBucketStep.class);
+  private final AzureConfiguration azureConfig;
+  private final CrlService crlService;
+  private final ControlledAzureIpResource resource;
+  private final WorkspaceService workspaceService;
+  private final AzureCloudContext azureCloudContext;
+
+  public CreateAzureIpStep(
+      AzureConfiguration azureConfig,
+      AzureCloudContext azureCloudContext,
+      CrlService crlService,
+      ControlledAzureIpResource resource,
+      WorkspaceService workspaceService) {
+    this.azureConfig = azureConfig;
+    this.azureCloudContext = azureCloudContext;
+    this.crlService = crlService;
+    this.resource = resource;
+    this.workspaceService = workspaceService;
+  }
+
+  @Override
+  public StepResult doStep(FlightContext context) throws InterruptedException, RetryException {
+    FlightMap inputMap = context.getInputParameters();
+    ApiAzureIpCreationParameters creationParameters =
+        inputMap.get(CREATION_PARAMETERS, ApiAzureIpCreationParameters.class);
+
+    ComputeManagerCow computeCow = crlService.getComputeManagerCow(azureCloudContext, azureConfig);
+
+    // Don't try to create it if it already exists. At this point the assumption is
+    // this is a redo and this step created it already.
+    PublicIpAddress existingIP;
+
+    try {
+      existingIP =
+          computeCow
+              .computeManager()
+              .networkManager()
+              .publicIpAddresses()
+              .getByResourceGroup(
```

Comment:
Instead of doing a get+create would it make sense to just `create` and recover from the conflict exception? That way we do 1 API call instead of 2.

Reply:
The azure API does not seem to return nice exceptions like google... the status code is embedded in a generic exception message body (see recovery block below). This is the pattern other components in WSM do (see create bucket), so I think its good for consistency. Not attached to this way though, if we can find an easily-parsable status code within the `com.azure.core.management.exception.ManagementException`